### PR TITLE
Add GLU variants

### PR DESCRIPTION
**Pull request by jaketae**
This PR addresses #44. Specifically, it implements the following activation functions:

* LiGLU
* GEGLU
* ReGLU
* SwiGLU

Nomenclature-wise, Bilinear seems to be synonymous with LiGLU.

### CODE REVIEW

**Comment by stas00** [comment]
Yes, of course! If it works that is.

**Comment by jaketae** [comment]
I think JIT doesn't like `None` as the default argument for bias. My understanding is that bias terms will be added in previous layers before the activation function, so I don't see why we would need it (I was following GPT-Neo's codebase as reference). I'll get rid of the bias argument and see if I can get JIT working.

**Comment by jaketae** [comment]
@stas00 Something funky is happening on my local, and I'm getting an error for all but `reglu`. Trace:

```
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
RuntimeError: vector
```

 I ran the code on Colab that verified that they all work fine. Should we be writing unit tests of any sort to ensure that the modules work as intended without error?

**Comment by stas00** [comment]
This looks like a truncated traceback, no? From googling this error typically there is `vector::something`

How do I reproduce it?

And yes, we absolutely need to start writing tests, as Meg didn't seem to have any. So if you're inspired start adding the tests under `tests` and overtime we will expand it and also need to setup up a CI.

**Comment by jaketae** [comment]
Surprisingly, that's all the trace shows me. Below is what I did to cause the error on my local (macOS, Python 3.7.3).

```python
>>> from megatron.model.activations import liglu, geglu, reglu, swiglu
>>> import torch
>>> x = torch.randn(8, 100, 768)
>>> reglu(x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jaketae/Documents/Dev/GitHub/Megatron-DeepSpeed/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
RuntimeError: vector

>>> torch.__version__
'1.9.0'
```

Two more observations:
* I was not able to reproduce the error on Colab
* When I moved `activations.py` to an entirely separate directly and ran the same sequence on the same virtual environment, the error did not occur

We will need testing to iron out messy details like this, and I could certainly try writing some very basic ones for the next few days, but unless you can reproduce the error on your local I wouldn't be too concerned... or should we?

**Comment by stas00** [comment]
I get the full trace on my machine:

```
python -c "from megatron.model.activations import liglu, geglu, reglu, swiglu; import torch; print(torch.__version__); reglu(torch.randn(8, 100, 768))"
2021-08-05 11:43:42.385851: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0
1.9.0
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/stas/anaconda3/envs/py38-pt19/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
RuntimeError: vector::_M_range_check: __n (which is 18446744073709551615) >= this->size() (which is 3)
```

File: megatron/model/activations.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+from torch.nn import functional as F
+
+
+class _GLUBaseModule(nn.Module):
+    def __init__(self, activation_fn):
+        super().__init__()
+        self.activation_fn = activation_fn
+    
+    def forward(self, x):
+        x1, x2 = x.chunk(2, dim=-1)
```
Comment (jaketae) [inline megatron/model/activations.py:12]:
```suggestion
        x1, x2 = x.chunk(2, dim=2)
```

**Comment by jaketae** [comment]
Apparently JIT doesn't like negative indexing ([source](https://discuss.pytorch.org/t/torch-jit-trace-unexpected-error-with-torch-cat-dim-1/55152)). I replaced `dim=-1` with `dim=2` and it works fine now (it also passed a mini unit test). Can you confirm? Thanks!

**Comment by stas00** [comment]
yeah, that fixed it. Interesting that dim=-1 is `18446744073709551615`

**Comment by stas00** [comment]
Is it always a dim 3 vector? It will fail with dim 4 vector for example

**Comment by jaketae** [comment]
It's odd that the error is still there, my source is from 2019. Is it safe to assume that all tensors will be three-dimensional?

**Comment by stas00** [comment]
yes, I just tested `x1, x2 = x.chunk(2, dim=(x.ndim-1))` as proposed in the thread you linked to and it works. Sounds like the most generic approach.

File: megatron/model/activations.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+from torch.nn import functional as F
+
+
+class _GLUBaseModule(nn.Module):
+    def __init__(self, activation_fn):
+        super().__init__()
+        self.activation_fn = activation_fn
+    
+    def forward(self, x):
+        x1, x2 = x.chunk(2, dim=2)
```
Comment (jaketae) [inline megatron/model/activations.py:12]:
```suggestion
        x1, x2 = x.chunk(2, dim=(x.ndim-1))
```

**Comment by stas00** [comment]
supposedly this bug has been fixed recently in https://github.com/pytorch/pytorch/issues/25135#event-4748477445, so probably pt-1.10.0, but we need to support older pytorch, so let's leave a comment there.

```
# dim=-1 breaks in pt<1.10
x1, x2 = x.chunk(2, dim=(x.ndim-1))
```

File: megatron/model/activations.py (hunk 2 above)
Comment (stas00) [reply megatron/model/activations.py:12]: edited to add a comment, looking good otherwise.

**Comment by jaketae** [comment]
Sounds good! Thanks for testing the code on your end.

File: megatron/model/activations.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+from torch.nn import functional as F
+
+
+class _GLUBaseModule(nn.Module):
+    def __init__(self, activation_fn):
+        super().__init__()
+        self.activation_fn = activation_fn
+    
+    def forward(self, x):
+        x1, x2 = x.chunk(2, dim=(x.ndim-1))
```
Comment (stas00) [inline megatron/model/activations.py:12]:
```suggestion
    def forward(self, x):
        # dim=-1 breaks in pt<1.10
        x1, x2 = x.chunk(2, dim=(x.ndim-1))
```

**Review by stas00 — APPROVED** [review]
Let's add basic tests, and looking great otherwise. Thank you, @jaketae

File: megatron/model/activations.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+from torch.nn import functional as F
+
+
+class _GLUBaseModule(nn.Module):
+    def __init__(self, activation_fn):
+        super().__init__()
+        self.activation_fn = activation_fn
+    
+    def forward(self, x):
+        # dim=-1 breaks in pt<1.10
```
Comment (stas00) [inline megatron/model/activations.py:12]:
```suggestion
        # dim=-1 breaks in jit for pt<1.10
```

sorry, I thought my initial comment was lacking important context that it's a jit issue. fixed now.

Comment (jaketae) [reply megatron/model/activations.py]: It's helpful detail, thanks for catching that.

**Comment by jaketae** [comment]
@stas00 I added some simple tests. I'm admittedly not used to writing test code and wasn't sure whether to go with monkey random tests or hard-coded inputs. I ended up largely sticking to the latter. Let me know if this is in line with what you had in mind. Thanks!

**Comment by jaketae** [comment]
@stas00 I wrote some simple tests. Admittedly I don't have a lot of experience writing test code, and I wasn't sure if the way I tested the operations make sense (i.e. the test somewhat regurgitates function definitions). I also took a monkey testing approach and used random inputs, with seeds set at the beginning of the run for reproducibility. Let me know what you think. Thanks!

**Comment by stas00** [comment]
Looks great, @jaketae 

We will gradually improve the test suite, so yours is a good start.

**Comment by jaketae** [comment]
Thanks for the feedback and review @stas00!

**Pull request merged**